### ISSUE TITLE
Why is a `'static` lifetime required for things that `impl FileSystem`?

### ISSUE THREAD

**Issue by landaire** [manuel-woelker/rust-vfs#66]
The core `FileSystem` trait is defined as:

```
pub trait FileSystem: Debug + Sync + Send + 'static
```

This prohibits someone from implementing this trait on something that requires borrowed data like so:

```rust
#[derive(Debug)]
struct StFS<'stfs, 'data> {
	package: &'stfs StfsPackage,
	data: &'data [u8],
}

impl<'stfs, 'data> FileSystem for StFS<'stfs, 'data> {
    // ...
}
```

From a cursory glance it seems like this is required because there's a `Box<dyn Filesystem>` involved with `VfsPath` -- which makes sense, but if I don't want to use `VfsPath` then this isn't necessarily a problem.

The following patch removes these restrictions:

```patch
diff --git a/src/filesystem.rs b/src/filesystem.rs
index 6124348..dce3e55 100644
--- a/src/filesystem.rs
+++ b/src/filesystem.rs
@@ -12,7 +12,7 @@ use std::time::SystemTime;
 /// Path components may be any UTF-8 string, except "/", "." and ".."
 ///
 /// Please use the test_macros [test_macros::test_vfs!] and [test_macros::test_vfs_readonly!]
-pub trait FileSystem: Debug + Sync + Send + 'static {
+pub trait FileSystem: Debug + Sync + Send {
     /// Iterates over all direct children of this directory path
     /// NOTE: the returned String items denote the local bare filenames, i.e. they should not contain "/" anywhere
     fn read_dir(&self, path: &str) -> VfsResult<Box<dyn Iterator<Item = String> + Send>>;
@@ -60,7 +60,7 @@ pub trait FileSystem: Debug + Sync + Send + 'static {
     }
 }
 
-impl<T: FileSystem> From<T> for VfsPath {
+impl<T: FileSystem + 'static> From<T> for VfsPath {
     fn from(filesystem: T) -> Self {
         VfsPath::new(filesystem)
     }
diff --git a/src/path.rs b/src/path.rs
index b6d7d45..7306223 100644
--- a/src/path.rs
+++ b/src/path.rs
@@ -141,7 +141,7 @@ impl VfsPath {
     /// # use vfs::{PhysicalFS, VfsPath};
     /// let path = VfsPath::new(PhysicalFS::new("."));
     /// ````
-    pub fn new<T: FileSystem>(filesystem: T) -> Self {
+    pub fn new<T: FileSystem + 'static>(filesystem: T) -> Self {
         VfsPath {
             path: "".into(),
             fs: Arc::new(VFS {
@@ -736,7 +736,7 @@ impl VfsPath {
     /// ```
     pub fn walk_dir(&self) -> VfsResult<WalkDirIterator> {
         Ok(WalkDirIterator {
-            inner: Box::new(self.read_dir()?),
+            inner: self.read_dir()?,
             todo: vec![],
         })
     }
```

Also note in `VfsPath::walk_dir()` I remove the `Box::new()` which isn't necessary, but the value was being double-boxed.

Would this patch be accepted or is there something I'm missing that makes lifting this restriction a bad idea?

**Comment by manuel-woelker** [comment]
Hi there, thanks for getting in touch.
The original idea was that all interaction with the FileSystem happens via `VfsPath` instances, and the filesystem is only used to construct the initial root `VfsPath`.
To ensure that this construction is possible, it made sense to require `FileSystem` to be `'static`.

From what you wrote, I assume you would like to use the "bare" FileSystem methods directly. Is that correct?
In that case removing the `'static` bound would probably be the way to go.

I was wondering if it would make sense to make the whole VfsPath generic over lifetimes.
Unfortunately, something like `struct VfsPath<'fs = 'static>` doesn't work since lifetime parameters cannot have default values.
Requiring lifetime annotations everywhere seems like quite a loss in usability.

Another option to mitigate that would be to type-alias `VfsPath` to `VfsPathWithLifetime<'static>`.

For your use case, wouldn't using VfsPaths for navigation etc. make more sense?
I also tried find the source for the `StfsPackage` to determine how it could be integrated, but I did not find anything.

**Comment by landaire** [comment]
>The original idea was that all interaction with the FileSystem happens via VfsPath instances, and the filesystem is only used to construct the initial root VfsPath.

My initial idea was to use this crate as a general abstraction to ensure I'm implementing everything that would be useful, but the more I've looked at it the more abundantly clear it's become that using a `VfsPath` as the main interface is a much better UX.

>I was wondering if it would make sense to make the whole VfsPath generic over lifetimes.

I took a stab at some refactoring that would allow this to be possible by making `VfsPath<T> where T: FileSystem`, and it worked for the most part (including moving files from `VfsPath<T>` to `VfsPath<U>`), but ran into troubles with `OverlayFS`.

My thought there was to refactor the core of `VfsPath` into the `PathLike` trait so I could `Box<dyn PathLike>` in the `OverlayFS`'s list of paths, but that also doesn't work because `VfsPath` functions return instances of `Self`, that can't be made into a trait, etc. etc.

>Requiring lifetime annotations everywhere seems like quite a loss in usability.

Agreed. I like the idea of the alias and can give that a try in my local branch. Early signal from a quick refactoring is that it might be more painful than initially thought (at least with my current patches).

You can find my WIP here: https://github.com/manuel-woelker/rust-vfs/compare/master...landaire:rust-vfs:master

The headache with trying to move forward will be trying to figure out a way to support a `VfsPath` that owns _or_ borrows its `FileSystem` without making `VfsPath` generic. Going to have to think that one over.

>For your use case, wouldn't using VfsPaths for navigation etc. make more sense?
>I also tried find the source for the StfsPackage to determine how it could be integrated, but I did not find anything.

Already answered that first part above but just to be explicit: I definitely think it'd make more sense and I was naive to think otherwise.

The `Stfs` source is part of this project that I'm currently refactoring. The WIP can be found here: https://github.com/landaire/acceleration/blob/dev/cli/src/main.rs


tl;dr of that project: the Xbox 360 has various filesystems for different kinds of data. For example, your player profile is in a content package on your hard disk that has a header (`XContentHeader`) describing the content and a signture over the content, and the content itself is an `StFS` (secure transactional(?) filesystem) package which implements a block device filesystem with data integrity features (hash trees + hashes over the blocks).

The files can be quite large, up to 4GB, so forcing the filesystem to own the data doesn't necessarily make sense.

In my case I'm `mmap()`ing the file and could easily `Arc<Mmap>` and store that in the filesystem, but I think that forcing the `FileSystem` to be constructed from an `Mmap` might be too strict.

**Comment by manuel-woelker** [comment]
Thanks for the background info! I took a quick look at https://free60.org/System-Software/Formats/STFS/ to better understand the format and limitations.

Re making VfsPath `VfsPath<FS: FileSystem>`: That was actually how this was defined in the first iteration of this library. Using it though I quickly found out that it has really cumbersome ergononmics, since now *every* operation dealing with IO (e.g. loading textures) had to be generic over `FS` (or box the paths). To make the developer experience smoother, I decided to box the filesystem, and rather have the virtual function call overhead (hopefully negligible when dealing with IO) rather than having too much generics and associated compile time overhead due to monomorphization.

Regarding implementing an StfsFilesystem: I assume that the actually files are stored on disk somewhere, and the primary use case is reading files. If that is the case, the StfsFileSystem would not have to "own" all the data in memory.
One approach might be to only parse the meta-information/header data and keep that in memory, since it should be relatively small.
When consumers read the files, the file-offsets in the header could be used to find the correct bytes on disk, and then read those on demand.

This would mean removing the lifetime bound from StfsPackage and duplicating/owning any of the currently shared/borrowed buffers. Since this should only be metadata, this would only be a small amount of data AFAICT. Overall I think this might make the access simpler, and allow `StfsFileSystem` to be `'static`. What are your thoughts on that approach?

Also, if you would like to make an independent PR for removing the double-boxing in `walk_dir()`, I'd be happy to merge that.

**Comment by landaire** [comment]
>When consumers read the files, the file-offsets in the header could be used to find the correct bytes on disk, and then read those on demand.

This is what I do today actually. I pre-calculate the block ranges that need to be read for each file, but the problem is that I'm not sure if it's appropriate for this structure to own the backing data for when it comes to read the bits... but perhaps that's a tradeoff that must be made.

>Also, if you would like to make an independent PR for removing the double-boxing in `walk_dir()`, I'd be happy to merge that.

SGTM!

**Comment by landaire** [comment]
FWIW I went with [`Arc<T> where T: AsRef<[u8]>`](https://github.com/landaire/acceleration/blob/d74adb870542e19165c62d5ae5fc892acd829a31/stfs/src/fs.rs#L20) for now.
